wallet: unify “allow/block other inputs“ concept

furszy opened the pull request:
Seeking to make the `CoinControl` options less confusing/redundant.
It should have no functional changes.

The too long to read technical description; remove `m_add_inputs`, we can use the already existent `fAllowOtherInputs` flag.

In #16377 the `CoinControl` flag ‘m_add_inputs’ was added to tell the coin filtering and selection process two things:
- Coin Filtering: Only use the provided inputs. Skip the Rest.
- Coin Selection: Search the wtxs-outputs and append all the `CoinControl` internal and external selected outpoints to the selection result (skipping all the available output checks). Nothing else.

Meanwhile, in `CoinControl` we already have a flag ‘fAllowOtherInputs’ which is already saying:
- Coin Filtering: Only use the provided inputs. Skip the Rest.
- Coin Selection: If false, no selection process -> append all the `CoinControl` selected outpoints to the selection result (while they passed all the `AvailableCoins` checks and are available in the 'vCoins' vector).

### Changes

As can notice, the first point in the coin filtering process is duplicated in the two option flags. And the second one, is slightly different merely because it takes into account whether the coin is on the `AvailableCoins` vector or not.
So it makes sense to merge ‘m_add_inputs’ and ‘fAllowOtherInputs’ into a single field for the coin filtering process while introduce other changes to add the missing/skipped internal and external coins into 'vCoins' vector if they were manually selected by the user.

——————————————————————————————————

Just as an extra note:
On top of this, I’m working on unifying/untangling further the coin filtering and selection processes so we have less duplicate functionality in both processes.

furszy committed:
wallet: unify “allow/block other inputs“ concept

Seeking to make the `CoinControl` option less confusing/redundant.

In #16377 the `CoinControl` flag ‘m_add_inputs’ was added to tell the coin filtering and selection process two things:
	- Coin Filtering: Only use the provided inputs. Skip the Rest.
	- Coin Selection: Search the wtxs-outputs and append all the `CoinControl` selected outpoints to the selection result (skipping all the available output checks). Nothing else.

Meanwhile, in `CoinControl` we already have a flag ‘fAllowOtherInputs’ which is already saying:
	- Coin Filtering: Only use the provided inputs. Skip the Rest.
	- Coin Selection: If false, no selection process -> append all the `CoinControl` selected outpoints to the selection result (while they passed all the `AvailableCoins` checks and are available in the 'vCoins' vector).

As can notice, the first point in the coin filtering process is duplicated in the two option flags. And the second one, is slightly different merely because it takes into account whether the coin is on the `AvailableCoins` vector or not.
So it makes sense to merge ‘m_add_inputs’ and ‘fAllowOtherInputs’ into a single field for the coin filtering process while introduce other changes to add the missing/skipped coins into 'vCoins' vector if they were manually selected by the user (follow-up commits).

achow101 commented:
Actually I don't think they are different behavior. They just are set and checked at different times so these two variables appear to be controlling different things.

If `m_add_inputs == false`, then we only filter for the preset inputs in `AvailableCoins`. If `fAllowOtherInputs == false`, we also only filter for the preset inputs in `AvailableCoins`. In `SelectCoins`, if `fAllowOtherInputs == false`, we have an optimization where we just select all of the preset inputs and exit early. For `m_add_inputs == false`, we don't do the optimization but instead select all of the inputs, do the coin selection algorithms, and arrive at the same selection of all of the preset inputs. So these two options fundamentally do the same thing and could actually be unified.

There is some weird interactions between then, such as when `m_add_inputs == false`, but `fAllowOtherInputs` is force set to `true` in `FundTransaction` - in this case `m_add_inputs` "overrides" `fAllowOtherInputs`, but it just isn't as obvious that it does. This would definitely be cleared up if they were the same variable.

furszy commented:
Hey achow101, thanks for the review!

> If m_add_inputs == false, then we only filter for the preset inputs in AvailableCoins. If fAllowOtherInputs == false, we also only filter for the preset inputs in AvailableCoins. In SelectCoins, if fAllowOtherInputs == false, we have an optimization where we just select all of the preset inputs and exit early. For m_add_inputs == false, we don't do the optimization but instead select all of the inputs, do the coin selection algorithms, and arrive at the same selection of all of the preset inputs. So these two options fundamentally do the same thing and could actually be unified.

I came to the same conclusion but.. the reason behind the new flag are all the `AvailableCoins` checks prior appending the coins to the `vCoins` vector that are being skipped in the second part of the `SelectCoins` process (where it loops over the `CoinControl` manually selected outputs and forcibly adds them to the tx, getting them from the wallet tx map directly instead of looking them inside `vCoins`..).

In other words, the usage of `m_add_inputs=true` was telling the wallet: "add this outputs as inputs into the tx; don't care if they are spent, locked or whatever state they are. Just add them".
While `HasSelected() + fAllowOtherInputs=false` is a "add this outputs as inputs into the tx only if they are in a valid state, not locked, with enough depth, etc".

An easy example of this is the locked coins skip that is inside `AvailableCoins` but not in the second part of `SelectCoins`:

The locked coins are, forcibly, selected inside the second part of `SelectCoins` while they were skipped inside `AvailableCoins`.. so if I change the current behavior (which I agree that is completely redundant: further notes below) and use the first part of the `SelectCoins` process: as the locked coins aren't inside `vCoins`, the process will not be able to select them, throwing an "Insufficient funds" error.

To try it, you can just set the new flag `m_use_manual_selection` to false at line 609 of `rpc/spend.cpp`, which without something like https://github.com/furszy/bitcoin-core/commit/1291ba7d2463c062242b82df879fba8ca29ba627 will fail at line 140 of the `rpc_psbt.py` test (which is the test line that checks that coin locks are ignored for manually selected inputs).

So.. conclusion:

 I agree with you and have been working on it on a local branch. Unifying and cleaning most of these redundancies (thus why found #25005). And added the manual flag here in order to not add a bulk of changes that are required to keep the same current behavior (explained above) without entering into the rabbit hole.

Getting deeper, some of the rabbit hole changes, linked to this work, that I have been working on locally (WIP) are:

1) The `SelectCoins` process shouldn't be searching coins in the wallet. That is exactly what `AvailableCoins` is already doing --> It only needs to be focused on the coins selection algorithm.
2) The wallet shouldn't be looping across the entire wallet's tx map if the user wants to select an specific set of inputs and nothing else --> it can just select them right away (skipping all the `AvailableCoins` checks using the manual selection flag).
3) The locked coins skip/add can be customizable by the user at selection time.

achow101 commented:
> In other words, the usage of `m_add_inputs=true` was telling the wallet: "add this outputs as inputs into the tx; don't care if they are spent, locked or whatever state they are. Just add them".
> While `HasSelected() + fAllowOtherInputs=false` is a "add this outputs as inputs into the tx only if they are in a valid state, not locked, with enough depth, etc".

I don't think this is a useful distinction.

The only ways to manually set inputs are through `send`, `fundrawtransaction`, `walletcreatefundedpsbt`, and the GUI. Those 3 RPCs all use the same `FundTransaction` function which currently does `fAllowOtherInputs = true`, so they always add the specified inputs regardless of state. The GUI only allows users to preset input that are shown in a list - a list which is created by doing `AvailableCoins`. So those inputs already only consists of coins that are in a valid state, and hence any additional checks for those valid states are redundant. So `HasSelected + fAllowOtherInputs=false` is not materially useful and we could just unify both parameters to what `m_add_inputs=false` does (with the optimization to skip running the selection algos) and users would not observe any difference.

achow101 commented:
Okay, I think we're on the same page now.

I think it's actually sufficient to drop the iteration of vCoins in the `HasSelected + fAllowOtherInputs = true` if-block (https://github.com/bitcoin/bitcoin/blob/629e250cbdee84c20d362da845d7aacfb84ddabe/src/wallet/spend.cpp#L429L444) and move it to after we go through `vPresetInputs` (after https://github.com/bitcoin/bitcoin/blob/629e250cbdee84c20d362da845d7aacfb84ddabe/src/wallet/spend.cpp#L447L492). This would mean that we will add all pre-selected inputs regardless of spent or locked, and do all of the necessary lookups to compute their sizes and fees. This would obviate the need for additional changes to `AvailableCoins`.

furszy committed:
wallet: move "use-only coinControl inputs" below the selected inputs lookup

Otherwise, RPC commands such as `walletcreatefundedpsbt` will not support the manual selection of locked, spent and externally added coins.

Full explanation is inside #25118 comments but brief summary is:

`vCoins` at `SelectCoins` time could not be containing the manually selected input because, even when they were selected by the user, the current `AvailableCoins` flow skips locked and spent coins.

Extra note: this is an intermediate step to unify the `fAllowOtherInputs`/`m_add_inputs` concepts. It will not be a problem anymore in the future when we finally decouple the wtx-outputs lookup process from `SelectCoins` and don't skip the user's manually selected coins in `AvailableCoins`.

furszy committed:
refactor: use GetWalletTx in SelectCoins instead of access mapWallet

achow101 commented on src/wallet/coincontrol.h:
```diff
@@ -33,11 +33,10 @@ class CCoinControl
-    //! If false, allows unselected inputs, but requires all selected inputs be used
+    //! If true, the selection process can add extra unselected inputs from the wallet
+    //! while requires all selected inputs be used
     bool fAllowOtherInputs = false;
```
nit: It would be nice if we could switch to using the new naming scheme. Perhaps a scripted diff?

furszy replied:
pushed

achow101 reviewed (commented):
ACK 4be088798dbb74380e6b07fb87569d83a1d02f1d

furszy committed:
scripted-diff: rename fAllowOtherInputs -> m_allow_other_inputs

-BEGIN VERIFY SCRIPT-
sed -i 's/fAllowOtherInputs/m_allow_other_inputs/g' -- $(git grep --files-with-matches 'fAllowOtherInputs')
-END VERIFY SCRIPT-

furszy commented:
hey @Sjors (sorry for the ping, didn't find you on IRC), you might be interested on this.

laanwj commented:
i get a build error with this merged to master:
```
…/bitcoin/src/wallet/test/coinselector_tests.cpp: In member function ‘void wallet::coinselector_tests::bnb_search_test::test_method()’:
…/bitcoin/src/wallet/test/coinselector_tests.cpp:395:22: error: ‘class wallet::CCoinControl’ has no member named ‘fAllowOtherInputs’; did you mean ‘m_allow_other_inputs’?
  395 |         coin_control.fAllowOtherInputs = true;
      |                      ^~~~~~~~~~~~~~~~~
      |                      m_allow_other_inputs
```

furszy commented:
great, thanks @laanwj 👌🏼. Rebased on master, silent issue fixed.

furszy commented:
rebased, conflicts solved. Ready to go.

laanwj commented:
Code review ACK d33871288636d9e4cdde2e1430de1f4223e2dd3f

beirut-boop commented:
Noticing some new irrational behavior in a forked coin control and this is the most significant work on coin control pulled from upstream. @furszy any change you could glance over the screenshots and assess the probability of this being an upstream issue?

https://github.com/syscoin/syscoin/issues/524

fanquake commented:
Sorry, this is not a support forum for forks of this project.

furszy commented:
> @fanquake I understand but the implication would be that there could be a bug in this project. If @furszy had responded differently after a quick glance, my next step would be to create a test case using Bitcoin Qt.

What fanquake pointed out, which I agree, is about linking other repositories issues here. Which isn't the best. We don't know what you have there (nor people are going to get deeper trying to find it). Time is a scarce resource for everyone.

But happy to help if you create an issue running bitcoin-core and describing your problematic next time.

beirut-boop commented:
@furszy I understand, concur, and have removed the link from my comment.

> Time is a scarce resource for everyone.

I agree. That's why I thought a quick assessment on the observed issue with the coin control classes before spending time setting up a Bitcoin testnet wallet with the intention of taking up more of your time to look into it was a good approach. The project is different but the coin control classes are identical, so I assumed I was helping everyone.

Apologies. I will not cross contaminate the comment section again in the future. :+1:
